### DISCUSSION TITLE
Best practice for bb script dev?

### DISCUSSION THREAD

**Discussion by rlhk** [babashka/babashka#1102] [Q&A]
I'm working on a utility bb script (those start with `#!/usr/bin/env bb`). Going for a proper project with ns seems to be an overkill. I'd like to properly test my functions, though less than 10. Wondering what's the best practice for TDD setup? Say run a watcher to reload on any code change, until all tests pass. I read through the official book and didn't find anything relevant. Did I miss anything?

Thanks!

**Comment by borkdude** [comment]
Something like this might work:

``` clojure
(ns script
  (:require [clojure.test :as t :refer [with-test is]]))

(with-test
  (defn foo
    []
    :foobar)
  (is (= :foobar (foo))))

(t/run-tests *ns*)
```

As for the filewatcher, you can use https://github.com/babashka/pod-babashka-fswatcher to run tests when files have changed:

``` clojure
(ns script
  (:require [clojure.test :as t :refer [with-test is]]))

(with-test
  (defn foo
    []
    :foobar)
  (is (= :foobar (foo))))

(require '[babashka.pods :as pods])
(pods/load-pod 'org.babashka/fswatcher "0.0.2")
(require '[pod.babashka.fswatcher :as fw])

(defmacro when-dev [& body]
  (when (System/getenv "DEV")
    `(do ~@body)))

(defn run-tests []
  (t/run-tests *ns*))

(when-dev
    (defonce watcher (fw/watch (str *file*)
                               (fn [event]
                                 (prn event)
                                 (case (:type event)
                                   (:write :chmod|write)
                                   (do (load-file *file*) (run-tests))
                                   nil)))))

(when-dev
    ;; in development, wait while watching script for running tests
    (run-tests)
    @(promise)
   )
```

**Reply by rlhk** [reply]
Thanks for moving the question from issue to discussion.

I've adapted the suggestion to use `#!/usr/bin/env bb` instead of `(ns ...)`, which has same behaviour as below. 
However, I found that although watcher reacts with outputs, both versions don't actually pickup any source changes. No matter inside or outside `with-test`. Looks like tests are always run against the initial code. What could be wrong? I'm on macOS.

```
#!/usr/bin/env bb
; Do `export DEV=true` for dev
(require '[clojure.test :as t :refer [deftest with-test is]])
(require '[babashka.pods :as pods])
(pods/load-pod 'org.babashka/fswatcher "0.0.2")
(require '[pod.babashka.fswatcher :as fw])

(defn bar [] :bar)
(defn foo [] :foo)

(deftest my-test
  (is (= :foo (foo)))
  (is (= :bar (bar))))

(defmacro when-dev [& body]
  (when (System/getenv "DEV")
    `(do ~@body)))

(defn run-tests []
  (t/run-tests *ns*))

(when-dev
  (defonce watcher (fw/watch (str *file*)
                             (fn [event]
                               (prn event)
                               (case (:type event)
                                 (:write :chmod|write)
                                 (run-tests)
                                 nil)))))

(when-dev
  ;; in development, wait while watching script for running tests
  (run-tests)
  @(promise))
```

**Reply by borkdude** [reply]
Here is an improved version:

``` clojure
#!/usr/bin/env bb
;; Do `export DEV=true` for dev
(ns example
  (:require [clojure.test :as t :refer [deftest is]]))

(defn bar [] :bar)
(defn foo [] :foo)

(deftest my-test
  (is (= :foo (foo)))
  (is (= :bar (bar))))

(defn run-tests []
  (t/run-tests *ns*))

(defmacro when-dev [& body]
  (when (System/getenv "DEV")
    `(do ~@body)))

#_{:clj-kondo/ignore [:unresolved-namespace]}
(when-dev
    (require '[babashka.pods :as pods])
    (pods/load-pod 'org.babashka/fswatcher "0.0.2")
    (require '[pod.babashka.fswatcher :as fw])
    (run-tests)
    (defonce watcher (fw/watch (str *file*)
                               (fn [event]
                                 (prn event)
                                 (case (:type event)
                                   (:write :chmod|write)
                                   (do (load-file *file*)
                                       (run-tests))
                                   nil))))
    @(promise))
``` 

I added `(load-file *file*)` so it evaluates the code of the current script.

**Reply by rlhk** [reply]
The improved version picks up code change for only once, then no reaction to change afterwards.
However it works as expected when I change `(defonce watcher ...` to `(def watcher ...`

**Reply by borkdude** [reply]
Changing it to `def` will create a new watcher every time the file reloads so you and up with multiple watcher, so that's no good. But you get the idea now I hope :).

**Reply by rlhk** [reply]
By the way, if I name the script as `test_example.clj` and namespace as `(ns test-example ...`, same one-time pickup behaviour occurs. The script only works as expected if I use a simple filename say `example.clj`. Whether I use one of the following require forms seems doesn't matter.
```
(ns example
   (:require [clojure.test :as t :refer [deftest is]]))
```
or `(require '[clojure.test :as t :refer [deftest is]])`

Perhaps an additional precaution to note?

**Reply by rlhk** [reply]
> Changing it to `def` will create a new watcher every time the file reloads so you and up with multiple watcher, so that's no good. But you get the idea now I hope :).

Yeah, got the idea. Thanks!
And there's a bug that complicated my verification. The case part should be:
```
(case (:type event)
  :write|chmod
  ...
```
Not sure if event keyword differs on different OS?

That said, I could only get reliable behaviour with `def` instead of `defonce` and the filename precaution.

**Reply by borkdude** [reply]
Maybe it's easier to write a watcher script that just reload another script when that script changes instead of having both the watcher and the watchee as one file.

**Reply by rlhk** [reply]
Yes, agree it should be easier if the main script is namespaced.

**Reply by rlhk** [reply]
For recording purpose: can confirmed that the said filename issue is a non-issue.
